### PR TITLE
[CARBONDATA-3874] segment mismatch between maintable and SI table when load with concurrency

### PR DESCRIPTION
**Pull request by maheshrajus**
segment mismatch between maintable and SI table when load with concurrency

 ### Why is this PR needed?
1) segment mismatch between main table and SI table with concurrency loads
2)  In concurrent loads, if one of the load failed for SI table then 'isSITableEnabled' will be disabled(isSITableEnabled = false).
    So in failed SI event listener case we are just checking  SI enabled is true(isSITableEnabled == true) then we are not loading current load to SI table. In concurrent scenarios, this might be happening as SI enabled state is true but segment difference may exist.

 ### What changes were proposed in this PR?
So instead of checking just  SI enabled is true(isSITableEnabled == true) we should also check if any segment difference between main table and SI table. The final output flag checking will be as follows.
``
   if ( isSITableEnabled == true || mainTblAndSidiff == true ) {
      ---
    }
``

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

https://issues.apache.org/jira/browse/CARBONDATA-3874

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1500/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3228/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 1]
```diff
@@ -390,8 +391,15 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
       }
       val indexTablePath = CarbonTablePath
         .getMetadataPath(tableInfo.getOrCreateAbsoluteTableIdentifier.getTablePath)
+      val metaPathMainTbl: String = carbonTable.getMetadataPath
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:394]: If metaPathMainTbl and  metaPathSI variable used in one place only, can replace with  `SegmentStatusManager.readLoadMetadata(carbonTable.getMetadataPath)`

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala [hunk 2]
```diff
@@ -78,20 +78,29 @@ class SILoadEventListenerForFailedSegments extends OperationEventListener with L
...
                   .getOrElse("isSITableEnabled", "true").toBoolean
+                val indexTable = metaStore
+                  .lookupRelation(Some(carbonLoadModel.getDatabaseName), indexTableName)(
+                    sparkSession)
+                  .asInstanceOf[CarbonRelation]
+                  .carbonTable
 
-                if (!isLoadSIForFailedSegments) {
+                val metaPathMainTbl: String = carbonTable.getMetadataPath
+                val listOfLoadFolderDetailsArrayMainTbl: Array[LoadMetadataDetails] =
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala:88]: Handle Same as above comment

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala (hunk 1 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:394]: OK, fixed

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala (hunk 2 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala:88]: @Indhumathi27  Here we are preparing list of metadata path of main table and si table. So adding a new method for just getting the metadata path/list of metadatapaths is no use in every case.

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 3]
```diff
@@ -390,8 +391,13 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
       }
       val indexTablePath = CarbonTablePath
         .getMetadataPath(tableInfo.getOrCreateAbsoluteTableIdentifier.getTablePath)
+      val listOfLoadFolderDetailsArrayMainTbl: Array[LoadMetadataDetails] =
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:394]: `listOfLoadFolderDetailsArrayMainTbl` variable name is big and not clear, can give a meaningful name like `mainTblLoadMetadataDetails`

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 4]
```diff
@@ -390,8 +391,13 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
...
+        SegmentStatusManager.readLoadMetadata(carbonTable.getMetadataPath)
+      val listOfLoadFolderDetailsArraySI: Array[LoadMetadataDetails] =
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:396]: similar to above

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala [hunk 5]
```diff
@@ -78,20 +78,27 @@ class SILoadEventListenerForFailedSegments extends OperationEventListener with L
...
                     Some(carbonLoadModel.getDatabaseName))).storage.properties
                   .getOrElse("isSITableEnabled", "true").toBoolean
+                val indexTable = metaStore
+                  .lookupRelation(Some(carbonLoadModel.getDatabaseName), indexTableName)(
+                    sparkSession)
+                  .asInstanceOf[CarbonRelation]
+                  .carbonTable
 
-                if (!isLoadSIForFailedSegments) {
+                val listOfLoadFolderDetailsArrayMainTbl: Array[LoadMetadataDetails] =
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/events/SILoadEventListenerForFailedSegments.scala:87]: `listOfLoadFolderDetailsArrayMainTbl`  same as above comment

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java [hunk 6]
```diff
@@ -304,14 +304,16 @@ public static boolean updateLoadMetadataWithMergeStatus(CarbonTable indexCarbonT
 
   /**
    * Method to check if main table and SI have same number of valid segments or not
-   *
+   * @param mainTableLoadMetadataDetails
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java:307]: no need of this, just method explanation is enough, as variable name suggest what info is it.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1572/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3310/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3312/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1574/

**Comment by akashrn5** [comment]
LGTM

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3315/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1577/

**Comment by akashrn5** [comment]
LGTM

**Pull request closed without merging**